crates/re_sdk/src/binary_stream_sink.rs

```diff
@@ -0,0 +1,200 @@
+#[derive(Clone, Default)]
+struct BinaryStreamStorageInner(Arc<Mutex<std::io::Cursor<Vec<u8>>>>);
+
+impl std::io::Write for BinaryStreamStorageInner {
+    fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
+        self.0.lock().write(buf)
+    }
+
+    fn flush(&mut self) -> std::io::Result<()> {
+        self.0.lock().flush()
+    }
+}
+
+/// The storage used by [`BinaryStreamSink`].
+///
+/// Reading from this consumes the bytes from the stream.
+pub struct BinaryStreamStorage {
+    inner: BinaryStreamStorageInner,
+    rec: RecordingStream,
+}
+
+impl BinaryStreamStorage {
+    /// Create a new binary stream storage.
+    fn new(rec: RecordingStream) -> Self {
+        Self {
+            inner: Default::default(),
+            rec,
+        }
+    }
+
+    /// Read and consume the current contents of the buffer.
+    ///
+    /// This does not flush the underlying batcher.
+    /// Use [`BinaryStreamStorage::flush`] if you want to guarantee that all
+    /// logged messages have been written to the stream before you read them.
+    #[inline]
+    pub fn read(&self) -> Vec<u8> {
+        let mut buffer = std::io::Cursor::new(Vec::new());
+        std::mem::swap(&mut buffer, &mut *self.inner.0.lock());
+        buffer.into_inner()
+    }
+
+    /// Flush the batcher and log encoder to guarantee that all logged messages
+    /// have been written to the stream.
+    ///
+    /// This will block until the flush is complete.
+    #[inline]
+    pub fn flush(&self) {
+        self.rec.flush_blocking();
+    }
+}
+
+impl Drop for BinaryStreamStorage {
+    fn drop(&mut self) {
+        self.flush();
+        let bytes = self.read();
+
+        if !bytes.is_empty() {
+            re_log::warn!("Dropping data in BinaryStreamStorage");
+        }
+    }
+}
+
+/// Stream log messages to an in-memory binary stream.
+///
+/// The contents of this stream are encoded in the Rerun Record Data format (rrd).
+///
+/// This stream has no mechanism of limiting memory or creating back-pressure. If you do not
+/// read from it, it will buffer all messages that you have logged.
+pub struct BinaryStreamSink {
+    /// The sender to the encoder thread.
+    tx: Mutex<Sender<Option<Command>>>,
+
+    /// Handle to join the encoder thread on drop.
+    join_handle: Option<std::thread::JoinHandle<()>>,
+}
+
+impl Drop for BinaryStreamSink {
+    fn drop(&mut self) {
+        self.tx.lock().send(None).ok();
+        if let Some(join_handle) = self.join_handle.take() {
+            join_handle.join().ok();
+        }
+    }
+}
+
+impl BinaryStreamSink {
+    /// Create a pair of a new [`BinaryStreamSink`] and the associated [`BinaryStreamStorage`].
+    pub fn new(rec: RecordingStream) -> Result<(Self, BinaryStreamStorage), BinaryStreamSinkError> {
+        let storage = BinaryStreamStorage::new(rec);
+
+        // We always compress when writing to a stream
+        // TODO(jleibs): Make this configurable
```

Comment:
any harm in just exposing that as a flag on creation right away?

Reply:
Mostly that (1) it also applies to every other sink type and (2) I can't come up with a good reason that a user would ever set it to False.

I'd rather introduce them all together in a consistent way if and when  we have a compelling use-case.